modules/system/assets/js/framework.js

```diff
@@ -41,23 +41,36 @@ if (window.jQuery.request !== undefined) {
         $triggerEl.trigger(_event, context)
         if (_event.isDefaultPrevented()) return
 
-        var data = [$form.serialize()],
-            loading = options.loading !== undefined ? options.loading : null,
+        var loading = options.loading !== undefined ? options.loading : null,
             isRedirect = options.redirect !== undefined && options.redirect.length,
-            useFlash = options.flash !== undefined
+            useFlash = options.flash !== undefined,
+            sendFiles = $el.is(':file') || ($el[0] == $form[0] && $form.attr('enctype') == 'multipart/form-data') || options.sendFiles,
+            data = $triggerEl.find(':input').addBack(':input').serializeArray()
 
         $.each($el.parents('[data-request-data]').toArray().reverse(), function extendRequest() {
-            data.push($.param(paramToObj('data-request-data', $(this).data('request-data'))))
+            appendObjectToFormData(paramToObj('data-request-data', $(this).data('request-data')), data)
         })
 
-        if ($el.is(':input') && !$form.length) {
-            var inputName = $el.attr('name')
-            if (inputName !== undefined && options.data[inputName] === undefined)
-                options.data[inputName] = $el.val()
+        if (options.data !== undefined && !$.isEmptyObject(options.data)) {
+            appendObjectToFormData(options.data, data)
         }
 
-        if (options.data !== undefined && !$.isEmptyObject(options.data)) {
-            data.push($.param(options.data))
+        if (sendFiles) {
+            var formdata = new FormData()
+
+            $.each(data, function (index, field) {
+                formdata.append(field.name, field.value)
+            })
+
+            $triggerEl.find(':file').addBack(':file').each(function () {
+                var $el = $(this),
+                    name = $el.attr('name'),
+                    files = $el.prop('files')
+
+                $.each(files, function (index, file) {
+                    formdata.append(name, file)
+                })
+            })
```

Comment:
`serializeArray` doesn't serialize file inputs, so if we have requested them we need to add them to our AJAX data, which must be an instance of FormData. We simply append our current data to the `FormData` instance, we then find all the file inputs and add the files. This is an array of files, so we need to loop over this and append each one to the `FormData`.